Note on the use of the transfer ID value for purposes of temporal synchronization of message streams

pavel-kirienko opened the pull request:
Added a minor footnote on the page 28 section 4.2.1. [UAVCAN_Specification.pdf](https://github.com/UAVCAN/specification/files/2408965/UAVCAN_Specification.pdf)

pavel-kirienko commented:
Added two more clarifications that fix https://github.com/UAVCAN/specification/issues/8.

[UAVCAN_Specification.pdf](https://github.com/UAVCAN/specification/files/2408980/UAVCAN_Specification.pdf)

thirtytwobits commented on specification/transport_layer/transport_layer.tex:
```diff
@@ -379,7 +386,11 @@ \subsection{Redundant interface support}
-should be handled in the first order.
+should be handled in the first order\footnote{It should be kept in mind that redundant interfaces
+are used for increased fault tolerance, not for load sharing reasons.
+Whenever a node is connected to an interface, the likelihood of its failure is increased;
```
Comma splice and ambiguity in the predicate. Perhaps: Whenever a node is connected to an interface the likelihood of the interface failing is increased

thirtytwobits commented on specification/transport_layer/transport_layer.tex:
```diff
@@ -379,7 +386,11 @@ \subsection{Redundant interface support}
 If perfectly simultaneous frame submission is not possible, interfaces with lower numerical index
-should be handled in the first order.
+should be handled in the first order\footnote{It should be kept in mind that redundant interfaces
+are used for increased fault tolerance, not for load sharing reasons.
```
I think this first sentence is superfluous to the footnote and could be omitted.

kjetilkjeka replied:
I think including explicitly that "redundant interfaces are used for increased fault tolerance, not for load sharing reasons.". can clear up misunderstandings and should be included in the text itself and not as a footnote. 

I don't think "kept in mind" is a very good phrasing since this hasn't been mentioned in the text before.

thirtytwobits commented on specification/transport_layer/transport_layer.tex:
```diff
@@ -379,7 +386,11 @@ \subsection{Redundant interface support}
+should be handled in the first order\footnote{It should be kept in mind that redundant interfaces
+are used for increased fault tolerance, not for load sharing reasons.
+Whenever a node is connected to an interface, the likelihood of its failure is increased;
+therefore, backup interfaces should interconnect only mission-critical equipment.
```
This is a non-normative comment so we shouldn't use "should" here. Perhaps: This suggests that backup interfaces may only interconnect with mission-critical equipment.

I know there is a counter argument to be made in having asymmetrical redundant busses where it adds complexity because the testing matrix is increased (i.e. you have to prove that mission-critical component X can operate nominally on bus B where node Y,Z, and Q are no longer available). This even makes the wiring harnesses different per bus which is, perhaps, added complexity and does mean the bus capacitance has to be calculated differently. Considering there is very little that is not mission-critical on some drones (i.e. where your ESCs are the only thing on the bus) I wouldn't put too much weight on this footnote.

thirtytwobits commented on specification/transport_layer/transport_layer.tex:
```diff
@@ -534,6 +545,16 @@ \subsection{State variables}
     or the protocol stack can estimate this value automatically by analyzing the transfer intervals. \\
 \end{UAVCANSimpleTable}
 
+It might be useful to clarify why immediate switching between interfaces is not allowed,
```
This is voiced conversationally whereas the rest of the specification is in "shall/should" voice. Could we adopt a box format to put non-normative and/or expositional text like this? It's jarring to be reading a spec and having a human suddenly take over the text and start talking.

pavel-kirienko replied:
> It's jarring to be reading a spec and having a human suddenly take over the text and start talking.

A very good point. I suggest just keeping the style consistent throughout rather than adopting the box convention you suggested. Fixed this in my latest commit.

pavel-kirienko commented:
I think I have addressed your concerns. I understand that my English is seriously lacking; however, I expect that @hamishwillee will be able to help us out on this when the first draft is nearly finished.

thirtytwobits commented on specification/transport_layer/transport_layer.tex:
```diff
@@ -534,6 +545,15 @@ \subsection{State variables}
     or the protocol stack can estimate this value automatically by analyzing the transfer intervals. \\
 \end{UAVCANSimpleTable}
 
+The reason nodes are required to let the first interface time out before using the next one
```
One more English nit:

Nodes are required to let the first interface timeout before using the next one because the transfer ID field is expected to wrap around frequently...

thirtytwobits commented on specification/transport_layer/transport_layer.tex:
```diff
@@ -534,6 +545,15 @@ \subsection{State variables}
+If the latency of a backup interface relative to the primary interface exceeds 32 transfer intervals,
+and receiving nodes were to be allowed to switch between interfaces freely disregarding the timeout,
+the receiving node would skip the whole period of transfer ID (32 transfers will be lost).
+The problem would primarily affect low-priority transfers where large latencies are possible.
```
s/affect/effect/

pavel-kirienko replied:
According to https://www.grammarly.com/blog/affect-vs-effect/ that would be a mistake.

thirtytwobits commented on specification/transport_layer/transport_layer.tex:
```diff
@@ -534,6 +545,15 @@ \subsection{State variables}
+especially if the system contains both redundantly-interfaced and non-redundantly-interfaced nodes.
+If the latency of a backup interface relative to the primary interface exceeds 32 transfer intervals,
+and receiving nodes were to be allowed to switch between interfaces freely disregarding the timeout,
+the receiving node would skip the whole period of transfer ID (32 transfers will be lost).
```
...the whole period of transfer IDs...
